clients/drcachesim/scheduler/scheduler_impl.cpp

```diff
@@ -2292,6 +2292,8 @@ scheduler_impl_tmpl_t<RecordType, ReaderType>::set_cur_input(
         // straightforward).
         if (options_.mapping == sched_type_t::MAP_TO_ANY_OUTPUT && prev_input != input &&
             !inputs_[prev_input].at_eof) {
+            // This is unsafe if the caller holds this input lock: we disallow it.
+            assert(!caller_holds_cur_input_lock);
```

Comment:
Given that this is an internal invariant for callers of this code, an abort or assert rather than gracefully handling it seems reasonable.  Once we're agreed on that: wouldn't most asserts result in crashes or maybe hangs or something similarly bad in release build?  Is your argument that we should not use asserts at all and should always use release-build-aborting checks?  We could decide to go that route: but currently we're using asserts in many places in clients/drcachesim/ for internal invariants that do not depend on user input.  It's a tradeoff: like you say, asserts only fire during testing or if a user deliberately turns on debug build.  But I would say it's weird to put in a release-build-abort here when we have asserts in other places: it should be a deliberate policy for the whole code base.  We do run all our tests in debug build which is one reason we've been ok with asserts.  So I would say this should match the other sanity checks and remain an assert.

Reply:
Changed the comment to `// The caller should never hold the input lock for MAP_TO_ANY_OUTPUT.`